js/src/builtin/streams/ReadableStreamBYOBReader.cpp

```diff
@@ -6,43 +6,552 @@
-  // Steps 3-5 are presently unreachable.
-  return nullptr;
+  Rooted<ReadableStream*> unwrappedStream(
+      cx, UnwrapAndTypeCheckArgument<ReadableStream>(
+              cx, args, "ReadableStreamBYOBReader constructor", 0));
+  if (!unwrappedStream) {
+    return false;
+  }
+
+  Rooted<JSObject*> reader(
+      cx, CreateReadableStreamBYOBReader(cx, unwrappedStream,
+                                         ForAuthorCodeBool::Yes, proto));
+  if (!reader) {
+    return false;
+  }
+
+  args.rval().setObject(*reader);
+  return true;
+}
+
+/**
+ * Streams spec, 3.6.4.1 get closed
+ */
+static bool ReadableStreamBYOBReader_closed(JSContext* cx, unsigned argc,
+                                            Value* vp) {
+  CallArgs args = CallArgsFromVp(argc, vp);
+
+  // Step 1: If ! IsReadableStreamBYOBReader(this) is false, return a promise
+  //         rejected with a TypeError exception.
+  Rooted<ReadableStreamBYOBReader*> unwrappedReader(
+      cx,
+      UnwrapAndTypeCheckThis<ReadableStreamBYOBReader>(cx, args, "get closed"));
+  if (!unwrappedReader) {
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+
+  // Step 2: Return this.[[closedPromise]].
+  Rooted<JSObject*> closedPromise(cx, unwrappedReader->closedPromise());
+  if (!cx->compartment()->wrap(cx, &closedPromise)) {
+    return false;
+  }
+
+  args.rval().setObject(*closedPromise);
+  return true;
+}
+
+/**
+ * Streams spec, 3.6.4.2. cancel ( reason )
+ */
+static bool ReadableStreamBYOBReader_cancel(JSContext* cx, unsigned argc,
+                                            Value* vp) {
+  CallArgs args = CallArgsFromVp(argc, vp);
+
+  // Step 1: If ! IsReadableStreamBYOBReader(this) is false, return a promise
+  //         rejected with a TypeError exception.
+  Rooted<ReadableStreamBYOBReader*> unwrappedReader(
+      cx, UnwrapAndTypeCheckThis<ReadableStreamBYOBReader>(cx, args, "cancel"));
+  if (!unwrappedReader) {
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+
+  // Step 2: If this.[[ownerReadableStream]] is undefined, return a promise
+  //         rejected with a TypeError exception.
+  if (!unwrappedReader->hasStream()) {
+    JS_ReportErrorNumberASCII(cx, GetErrorMessage, nullptr,
+                              JSMSG_READABLESTREAMREADER_NOT_OWNED, "cancel");
+    return ReturnPromiseRejectedWithPendingError(cx, args);
+  }
+
+  // Step 3: Return ! ReadableStreamReaderGenericCancel(this, reason).
+  JSObject* cancelPromise =
+      ReadableStreamReaderGenericCancel(cx, unwrappedReader, args.get(0));
+  if (!cancelPromise) {
+    return false;
+  }
+  args.rval().setObject(*cancelPromise);
+  return true;
+}
+
+/**
+ * https://streams.spec.whatwg.org/#byob-reader-read
+ */
+static bool ReadableStreamBYOBReader_read(JSContext* cx, unsigned argc,
+                                          Value* vp) {
+  CallArgs args = CallArgsFromVp(argc, vp);
+
+  Handle<Value> view = args.get(0);
```

Comment:
Do we need to check that an argument was first supplied?

Reply:
The validations below validate that the `view` is the right type of object, so I believe these paths catch the error ok?